File: RB110/ruby_interview_problems/new_rb119_interview_study_problems/problem19.rb
=begin
Create a method that takes an array of integers as an argument and returns
the integer that appears an odd number of times. There will always be exactly
one such integer in the input array.

p odd_fellow([4]) == 4
p odd_fellow([7, 99, 7, 51, 99]) == 51
p odd_fellow([7, 99, 7, 51, 99, 7, 51]) == 7
p odd_fellow([25, 10, -6, 10, 25, 10, -6, 10, -6]) == -6
p odd_fellow([0, 0, 0]) == 0


Problem: Given an array of integers, return an integer

Rules:
- Return integer will appear an odd amount of times
- Every other integer will appear an even amount of times
- There is only 1 integer w/ occurence sum that is odd

Algorithm:
- ITERATE over the array of nums
  - for each number
    - if the number of times the current num appears in the array is odd (its count)
      - return that number
    - else move onto the next number

=end

def odd_fellow(numbers)
  numbers.each { |num| return num if numbers.count(num).odd? }
end


p odd_fellow([4]) == 4
p odd_fellow([7, 99, 7, 51, 99]) == 51
p odd_fellow([7, 99, 7, 51, 99, 7, 51]) == 7
p odd_fellow([25, 10, -6, 10, 25, 10, -6, 10, -6]) == -6
p odd_fellow([0, 0, 0]) == 0

File: LS215/Lesson3/problem19.js
// 2.  Number Problem:
//    Create a function that takes an integer as an argument and returns the sum of all prime numbers up to and including that integer. If the input is less than 2, return 0.



// javascript

function isPrime(num) {
  for (let factor = 2; factor < num; factor += 1) {
    if (num % factor === 0) return false;
  }

  return true;
}

function sumOfPrimes(number) {
  if (!Number.isInteger(number)) return null;
  if (number < 2) return 0;

  let primeTotal = 0;

  for (let num = 2; num <= number; num += 1) {
    if (isPrime(num)) primeTotal += num;
  }

  return primeTotal;
}
console.log(sumOfPrimes([]));
console.log(sumOfPrimes(true));
console.log(sumOfPrimes('1'));
console.log(sumOfPrimes(1.1));
console.log(sumOfPrimes(10) === 17); // 2 + 3 + 5 + 7 = 17
console.log(sumOfPrimes(5) === 10); // 2 + 3 + 5 = 10
console.log(sumOfPrimes(1) === 0);
console.log(sumOfPrimes(0) === 0);
console.log(sumOfPrimes(20) === 77); // 2 + 3 + 5 + 7 + 11 + 13 + 17 + 19 = 77

// Additional edge cases
console.log(sumOfPrimes(-5) === 0);
console.log(sumOfPrimes(2) === 2);
console.log(sumOfPrimes(3) === 5);
console.log(sumOfPrimes(100) === 1060);
console.log(sumOfPrimes(Infinity) === null);
console.log(sumOfPrimes(NaN) === null);
console.log(sumOfPrimes() === null);
console.log(sumOfPrimes(null) === null);
console.log(sumOfPrimes(undefined) === null);